File: src/main/java/com/hazza/leetcode/SegmentThree/No124BinaryTreeMaximumPathSum.java
package com.hazza.leetcode.SegmentThree;

import com.hazza.leetcode.datastructure.TreeNode;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/binary-tree-maximum-path-sum/description/
 * User: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 18-2-7
 * Time: 1:29 PM
 */
public class No124BinaryTreeMaximumPathSum {
    int max;

    public int maxPathSum(TreeNode root) {
        if (root == null) return 0;
        max = Integer.MIN_VALUE;
        treeDFS(root);
        return max;
    }

    private int treeDFS(TreeNode node) {
        if (node == null) return 0;
        int left = Math.max(0, treeDFS(node.left));
        int right = Math.max(0, treeDFS(node.right));
        max = Math.max(max, left + right + node.val);

        return Math.max(left, right) + node.val;
    }
}

File: src/main/python/leetcode_by_python/dfs/No124_Binary_Tree_Maximum_Path_Sum.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-11-19
import sys

from src.main.python.leetcode_by_python.data_structure.TreeNode import TreeNode


class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        def dfs(node):
            if not node:
                return 0
            left = max(0, dfs(node.left))
            right = max(0, dfs(node.right))
            self.cur_max = max(self.cur_max, left + right + node.val)

            return max(left, right) + node.val

        self.cur_max = -sys.maxsize
        if not root:
            return 0
        dfs(root)

        return self.cur_max
